### DISCUSSION TITLE
Volume and mixer control prob

### DISCUSSION THREAD

**Discussion by giggywithit** [dtcooper/raspotify#610] [Q&A] [answered]
Just upgraded from buster and everything works except for the volume and mixer

Setting the initial volume actually sets my "Digital" mixer to that setting, however, adjusting the volume on the spotify app does not have any effect on "Digital". The previous function used to adjust "Digital" remotely in the app.

The app does control some volume, just not the mixer in the Rpi

Clues?
I seem to recall there was a rewrite to the sound system?
What changed over the last 2 yrs in the mixer dpt?

Thx

**Comment by giggywithit** [comment]
Ok I have a tug of war with shairport-sync
Shairport wants to control the DIgital part of my mixer
Setting my initial volume in raspotify had no effect in actuality
Still have to figure out what to do about the loss of controling "Digital" in spotify because shairport sets it when it grabs it and then when I want to use raspotify, it's stuck at a low setting.
Argh!

**Reply by giggywithit** [reply]
Just want to check here and see if there is a way to control the mixer or if the volume is now all software 

This part of the config leads me to believe that there is a way to control the alsamixer

![Screenshot from 2023-03-04 09-59-01](https://user-images.githubusercontent.com/36465355/222913026-bd331193-204a-465d-8367-ce9188c23280.png)

**Comment by JasonLG1979** [comment]
Your better off just using software volume for both and not touching the hardware volume at all. I'm not sure about `shairport-sync` but the software volume in librespot is done with f64 precision and is objective superior to any hardware volume control. If you use hardware volume control it will be a constant tug of war and the volume levels will get de-synchronized.

**Reply by giggywithit** [reply]
The issue is that on an amplifier, having hardware volume wide open often results in hiss through the speakers while nothing is played back. I'm sure the soft vol is superior, but it doesn't resolve the muting characteristics of a hardware volume control. 

Is there a setting somewhere where I can get hardware back?

Thx tons.

PS: do you need help getting a rpi4?
I seem to have read that you killed one

**Reply by JasonLG1979** [reply]
> Is there a setting somewhere where I can get hardware back?

Sure. `librespot -h` prints out all of the options and flags. There are ***a lot*** so I only included a few in the config but you add whatever you need. They just need to be converted to environment variables.
  
From the librespot wiki:

"Any cli argument you can pass to librespot can also be passed as an environment variable by prefixing it with `LIBRESPOT_`, making it all caps and replacing any `-` with `_`. So an option `foo-bar` would become the environment variable `LIBRESPOT_FOO_BAR`."

> PS: do you need help getting a rpi4?
I seem to have read that you killed one


I did but I managed to get my hands on a Pi Zero v2 which is more than enough for Raspotify.

**Reply by giggywithit** [reply]
Thanks Jason

I added a few lines with no apparent change in performance
The mixer control does nothing and when I add the LIBRESPOT_MIXER="alsa" it no longer shows up as a connect device
Thanks for all the pointers

![Screenshot from 2023-03-06 02-56-28](https://user-images.githubusercontent.com/36465355/223051439-379d19f5-d0cf-452e-8186-0136fa09fd21.png)

![Screenshot from 2023-03-06 03-07-10](https://user-images.githubusercontent.com/36465355/223052743-7793f62b-0012-4264-b5a7-350291336f1d.png)

**Reply by JasonLG1979** [reply]
So that worked for you then?

**Reply by giggywithit** [reply]
No
It still uses softvol

**Reply by giggywithit** [reply]
If I add the LIBRESPOT_MIXER="alsa" then it doesn't show up as a spotify device and cannot play anything

**Comment by giggywithit** [comment]
No
Has no effect

**Comment by giggywithit** [comment, marked as answer]
Sorry for all the cluttered comments, I haven't used these forums in a while...
I had a similar string of issues with the prior methods of passing commands to librespot using raspotify a few years ago and appear to have found the solution. It's a dark area in this space to find the solution unless you play around a bit to track down the issues, path and methods. In this instance, everything must be enabled in order for it to work correctly. I now have the hardware mixer working on my DAC.

Please add a note if you can about this in the config so that when I forget again, it will remind me!
![Screenshot from 2023-03-06 11-01-58](https://user-images.githubusercontent.com/36465355/223170510-28858f6f-d4ab-441a-9d32-ea92b33c0161.png)
![Screenshot from 2023-03-06 11-26-35](https://user-images.githubusercontent.com/36465355/223170737-796ca53d-28cb-48bb-af24-1759423b7403.png)

**Reply by busterwolfeman** [reply]
Many many thanks. This issue has been driven crazy today. With your help it worked now :-)
